# Problema 1 - Laboratorio 1 Cálculo Numérico
## Resolución de la ecuación de segundo grado
##### Antonio Cabrera Landín 3ºMAIS
---

### La ecuación de segundo grado $ax^2 + bx + c = 0$ se resuelve de la forma ordinaria utilizando las fórmulas
$$x_1 = \frac{-b + \sqrt{b^2 - 4ac}}{2a}, \;\; x_2 = \frac{-b - \sqrt{b^2 - 4ac}}{2a}$$

#### 1. Pruebe que si se define
$$y_1 =  \frac{2c}{-b - \sqrt{b^2 - 4ac}}, \;\; y_2 = \frac{2c}{-b + \sqrt{b^2 - 4ac}}$$
#### entonces $x_1 = y_1$ y $x_2 = y_2$

$$\begin{split}
    & x_1 = y_1 \Longleftrightarrow \frac{-b + \sqrt{b^2 - 4ac}}{2a} = \frac{2c}{-b - \sqrt{b^2 - 4ac}} \Longleftrightarrow \frac{(-b + \sqrt{b^2 - 4ac})(-b - \sqrt{b^2 - 4ac})}{(2a)(2c)} = 1 \Longleftrightarrow \frac{(-b)^2 - (\sqrt{b^2 - 4ac})^2}{4ac} = 1 \Longleftrightarrow \frac{b^2 - |b^2 - 4ac|}{4ac} = 1\\
    & \text{si } b^2 > 4ac \implies \frac{(-b)^2 - (b^2 - 4ac)}{4ac} = 1 \Longleftrightarrow \frac{(-b)^2 - b^2 + 4ac)}{4ac} = 1 \Longleftrightarrow \frac{4ac}{4ac} = 1\\
    & x_2 = y_2 \Longleftrightarrow \frac{-b - \sqrt{b^2 - 4ac}}{2a} = \frac{2c}{-b + \sqrt{b^2 - 4ac}} \Longleftrightarrow \frac{(-b)^2 - (\sqrt{b^2 - 4ac})^2}{4ac} = 1 \Longleftrightarrow \frac{b^2 - |b^2 - 4ac|}{4ac} = 1\\
    & \text{si } b^2 > 4ac \implies \frac{(-b)^2 - (b^2 - 4ac)}{4ac} = 1 \Longleftrightarrow \frac{(-b)^2 - b^2 + 4ac)}{4ac} = 1
\end{split}$$

Solución: para raices no complejas se cumplen las igualdades

#### 2. Evalúe, utilizando Python, $x_1,x_2,y_1,y_2$ cuando $a=1,b=9^{12}$ y $c=-3$

Primero vamos a implementar las formulas de cada una de las variables, para ello necesitaremos la librería base de `math`

In [6]:
import math

Las $x$ las calculamos con la fórmula de Bashkara. Tenemos que comprobar además que la raiz no sea compleja y que el denominador no sea 0 

In [7]:
def get_x1(a, b, c):

    if b**2 - 4*a*c < 0:
        raise ValueError('No se puede calcular raíces negativas')

    if a == 0:
        raise ValueError('No se puede dividir por 0')

    return (-b + math.sqrt(b**2 - 4*a*c)) / (2*a)

def get_x2(a, b, c):
    
        if b**2 - 4*a*c < 0:
            raise ValueError('No se puede calcular raíces negativas')
    
        if a == 0:igualdades
            raise ValueError('No se puede dividir por 0')
    
        return (-b - math.sqrt(b**2 - 4*a*c)) / (2*a)

In [24]:
a=1
b=0
c=-64
print(f'({a})x² + ({b})x + ({c}) = 0 -> x_1:{get_x1(a, b, c)}, x_2:{get_x2(a, b, c)}')

(1)x² + (0)x + (-64) = 0 -> x_1:8.0, x_2:-8.0


Las $y$ las calculamos con la fórmula dada, como Python puede llegar a aproximar el denominador a $0$, en esos casos utilizaremos que $x_1 = y_1$ y $x_2 = y_2$

In [12]:
def get_y1(a, b, c):
    
        if b**2 - 4*a*c < 0:
            raise ValueError('No se puede calcular raíces negativas')
        try: 
            result = (2*c) / (-b - math.sqrt(b**2 - 4*a*c))
        except ZeroDivisionError:
            print('y1: usando la fórmula de x1')
            result = get_x1(a, b, c)

        return result

def get_y2(a, b, c):

    if b**2 - 4*a*c < 0:
        raise ValueError('No se puede calcular raíces negativas')

    try:
        result = (2*c) / (-b + math.sqrt(b**2 - 4*a*c))
    except ZeroDivisionError:
        print('y2: usando la fórmula de x2')
        result = get_x2(a, b, c)

    return result

In [23]:
a=1
b=0
c=-64
print(f'({a})x² + ({b})x + ({c}) = 0 -> y_1:{get_y1(a, b, c)}, y_2:{get_y2(a, b, c)}')

(1)x² + (0)x + (-64) = 0 -> y_1:8.0, y_2:-8.0


Podemos comprobar que efectivamente $x_1 = y_1$ y $x_2 = y_2$

In [17]:
a=1
b=0
c=-64
print(f'Es x_1 = y_1: {get_x1(a, b, c) == get_y1(a,b,c)}')
print(f'Es x_2 = y_2: {get_x2(a, b, c) == get_y2(a, b, c)}')

Es x_1 = y_1: True
Es x_2 = y_2: True


Ahora ya podemos calcular $x_1,x_2,y_1,y_2$ para $a=1,b=9^{12}$ y $c=-3$. Aunque a primera vista parece que $x_1 \neq y_1$, esto ocurre por el redondeo decimal de Python

In [26]:
a=1
b=9**12
c=-3
print(f'({a})x² + ({b})x + ({c}) = 0 -> x_1:{get_x1(a, b, c)}, x_2:{get_x2(a, b, c)}')
print(f'({a})x² + ({b})x + ({c}) = 0 -> y_1:{get_y1(a, b, c)}, y_2:{get_y2(a, b, c)}')

(1)x² + (282429536481)x + (-3) = 0 -> x_1:0.0, x_2:-282429536481.0
y2: usando la fórmula de x2
(1)x² + (282429536481)x + (-3) = 0 -> y_1:1.062211848441645e-11, y_2:-282429536481.0


#### 3. Evalúe, utilizando Python, $x_1,x_2,y_1,y_2$ cuando $a=1,b=-9^{12}$ y $c=-3$

In [25]:
a=1
b=-9**12
c=-3
print(f'({a})x² + ({b})x + ({c}) = 0 -> x_1:{get_x1(a, b, c)}, x_2:{get_x2(a, b, c)}')
print(f'({a})x² + ({b})x + ({c}) = 0 -> y_1:{get_y1(a, b, c)}, y_2:{get_y2(a, b, c)}')

(1)x² + (-282429536481)x + (-3) = 0 -> x_1:282429536481.0, x_2:0.0
y1: usando la fórmula de x1
(1)x² + (-282429536481)x + (-3) = 0 -> y_1:282429536481.0, y_2:-1.062211848441645e-11


#### 4. Para cada uno de los dos apartados anteriores, indica cuáles son las soluciones de la ecuación de segundo grado y explique los resultados obtenidos en python

- En el apartado 2, la solución a la ecuación $x^2 + 9^{12}x - 3 = 0$ es $x_1 = 1.06 \cdot 10^{-11}$ y $x_2 = -282429536481$
- En el apartado 3, la solución a la ecuación $x^2 - 9^{12}x - 3 = 0$ es $x_1 = 282429536481$ y $x_2 = -1.06 \cdot 10^{-11}$

Como ya hemos visto anteriormente, en python obtenemos ceros en las $x$, esto se debe a que python redondea los valores que son menores de $2.220446049250313 \cdot 10^{-16}$